File: py/tokenizer.py
# -*- coding: utf-8 -*-
# @Author: Jingyuexing
# @Date:   2024-05-05 09:23:29
# @Last Modified by:   Jingyuexing
# @Last Modified time: 2024-05-06 23:21:26
from enum import Enum

class TokenKind(Enum):
    GreatThan = "GreatThan"
    LessThan = "LessThan"
    LessThanOrEqual = "LessThanOrEqual"
    GreatThanOrEqual = "GreatThanOrEqual"
    COLON = "Colon"
    Number = "Number"
    Text = "Text"
    Identifier = "Identifier"
    NotEqual = "NotEqual"
    Comma = "Comma"
    Terminator = "Terminator"

class Token:
    def __init__(self, type:'TokenKind', value:str):
        self.type = type
        self.value = value

def create_token(type:"TokenKind", value:str):
    return Token(type, value)

def is_numeric(ch):
    return "0" <= ch <= "9"

def is_letter(ch):
    return ("a" <= ch <= "z") or ("A" <= ch <= "Z") or ch >= "\xff"

def is_whitespace(ch):
    return ch == "\n" or ch == " " or ch == "\r" or ch == "\t"

def tokenizer(text):
    current = 0
    tokens = []
    
    def advance():
        nonlocal current
        current += 1
    
    while current < len(text):
        ch = text[current]
        
        if is_whitespace(ch):
            while is_whitespace(ch) and current < len(text):
                advance()
                ch = text[current]
            tokens.append(create_token(TokenKind.Terminator, ";"))
            continue
        
        if is_numeric(ch):
            value = ""
            while (is_numeric(ch) or ch == ".") and current < len(text):
                value += ch
                advance()
                ch = text[current]
            tokens.append(create_token(TokenKind.Number, value))
        
        if is_letter(ch):
            value = ""
            while (is_letter(ch) or is_numeric(ch) or ch in ["-", "$", "_"]) and current < len(text):
                value += ch
                advance()
                ch = text[current]
            tokens.append(create_token(TokenKind.Identifier, value))
        
        if ch == ":":
            tokens.append(create_token(TokenKind.COLON, ch))
            advance()
            continue
        if ch == ",":
            tokens.append(create_token(TokenKind.Comma,ch))
            advance()
            continue
        if ch == "<":
            slice_ = text[current: current + 2]
            if slice_ == "<=":
                tokens.append(create_token(TokenKind.LessThanOrEqual, slice_))
                current += 2
            elif slice_ == "<>":
                tokens.append(create_token(TokenKind.NotEqual, slice_))
                current += 2
            else:
                tokens.append(create_token(TokenKind.LessThan, ch))
                advance()
        
        if ch == ">":
            slice_ = text[current: current + 2]
            if slice_ == ">=":
                tokens.append(create_token(TokenKind.GreatThanOrEqual, slice_))
                current += 2
            else:
                tokens.append(create_token(TokenKind.GreatThan, ch))
                advance()
    
    return tokens

File: ts/src/tokenizer.ts
function isNumberic(ch: string) {
	return "0" <= ch && ch <= "9";
}
function isLetter(ch: string) {
	return ("a" <= ch && ch <= "z") || ("A" <= ch && ch <= "Z") || ch >= "\xff";
}

function isWhiteSpace(ch: string) {
	return "\n" === ch || " " === ch || ch === "\r" || ch === "\t";
}

export enum TokenKind {
	GreatThan = "GreatThan",
	LessThan = "LessThan",
	LessThanOrEqual = "LessThanOrEqual",
	GreatThanOrEqual = "GreatThanOrEqual",
	COLON = "Colon",
	Number = "Number",
	Text = "Text",
	Identifier = "Identifier",
	NotEqual = "NotEqual",
	Terminator = "Terminator",
	Comma = "Comma"
}
export interface Token {
	type: TokenKind;
	value: string;
}

function createToken(type: TokenKind, value: string) {
	return {
		type,
		value,
	};
}

// < =
// > =
// < >
// :
// repo:name

export function tokenizer(text: string) {
	let current = 0;
	const tokens: Token[] = [];
	let ch = text[current];
	function adavance() {
		current++;
		ch = text[current];
	}
	while (current < text.length) {
		ch = text[current];
		if (isWhiteSpace(ch)) {
			while(isWhiteSpace(ch) && current < text.length){
				adavance();
			}
			tokens.push(createToken(TokenKind.Terminator,";"))
			continue;
		}
		if (isNumberic(ch)) {
			let value = "";
			while ((isNumberic(ch) || ch === ".") && current <= text.length) {
				value += ch;
				adavance();
			}
			tokens.push(createToken(TokenKind.Number, value));
		}
		if (isLetter(ch)) {
			let value = "";
			while ((isLetter(ch) || ["-","$","_"].includes(ch)) && current <= text.length) {
				value += ch;
				adavance();
			}
			tokens.push(createToken(TokenKind.Identifier, value));
		}
		if (ch === ":") {
			tokens.push(createToken(TokenKind.COLON, ch));
			adavance()
			continue;
		}
		if(ch === ","){
			tokens.push(createToken(TokenKind.Comma,ch))
			adavance()
			continue;
		}
		if (ch === "<") {
			const slice = text.slice(current, current + 2);
			switch (slice) {
				case "<=":
					tokens.push(createToken(TokenKind.LessThanOrEqual, slice));
					current += 2;
					break;
				case "<>":
					tokens.push(createToken(TokenKind.NotEqual, slice));
					current += 2;
					break;
				default:
					tokens.push(createToken(TokenKind.LessThan, ch));
					adavance();
			}
		}
		if (ch === ">") {
			const slice = text.slice(current, current + 2);
			switch (slice) {
				case ">=":
					tokens.push(createToken(TokenKind.GreatThanOrEqual, slice));
					current +=2
					break;
				default:
					tokens.push(createToken(TokenKind.GreatThan, ch));
					adavance();
			}
		}
	}
	return tokens;
}